File: tools/crypto_bear_recovery.py
#!/usr/bin/env python3
# =============================================================================
# crypto_bear_recovery.py -- BearRecovery long-only signal engine for BTC/ETH
# (S-2026-07-03, operator: "trade the bounces in the current downtrend, max
#  protection, max aggression, without impacting the edges if found").
#
# WHAT IT IS: the deployable half of backtest/crypto_bear_bounce/ (READ
# FINDINGS.md FIRST). It emits the full regime ladder + entry/exit/protection
# state per symbol as JSON, for the ~/Crypto book (stall_accountant /
# refresh_shadow cycle) or manual use. It does NOT place orders itself --
# execution stays with the book / omega_crypto_bridge.py (paper 4002).
#
# THE LADDER (daily closes, completed UTC bars only):
#   KNIFE     close < SMA200 and NOT (close > SMA50 and SMA50 rising)
#             -> FLAT. No long-only entry family survives here (7 families
#                backtested 2017-2026, all PF<1 in 2018/2022/2025-26 knives).
#                The protection IS the flatness.
#   RECOVERY  close < SMA200 AND close > SMA50 AND SMA50 > SMA50[5d ago]
#             -> this engine trades: LONG on daily close crossing above EMA9
#                after >=3 closes below. Ride to first daily close < EMA9.
#   BULL      close > SMA200 -> hand off to the Luke daily system (gate open).
#
# ADVERSE-PROTECTION: backtested verdict (FINDINGS.md sect.4-5, 2026-07-03) --
#   initial stop = entry-day low - 0.5*ATR14(D) (intrabar, hard);
#   BE-AND-RIDE floor: after MFE >= +2%, floor = entry (lock 0); exit only
#     through the floor -- net UP, PF 5.16->8.85, worst -8.3%->-4.8%;
#   COLD CUT: no-op at all values (EMA9-flip already bounds never-green) -- OFF;
#   GIVEBACK clips: proven harmful (net -62..-90%) -- MUST NOT be added.
# Sizing: risk-based off the structural stop; ship 2% equity risk (3% = max
# aggressive tier; PF holds >=4.2 to 5%). ~2 signals/yr -- aggression comes
# from size, not frequency.
#
# Usage:
#   python3 crypto_bear_recovery.py                    # one-shot, both symbols
#   python3 crypto_bear_recovery.py --symbols BTC-USD  # subset
#   python3 crypto_bear_recovery.py --equity 50000 --risk-pct 0.02
# Output: one JSON line per symbol:
#   {"sym","regime","signal","entry_ref","stop","be_arm_pct","be_lock",
#    "qty_usd","exit_flag","asof","detail"}
#   signal: "ENTER" only on the day the reclaim cross fires; "HOLD-RULES" when
#   a position (if any) should keep running; "EXIT" when close < EMA9; "NONE".
# =============================================================================
import argparse, json, urllib.request, datetime as dt

CB = "https://api.exchange.coinbase.com/products/{sym}/candles?granularity=86400&start={s}&end={e}"
UA = {"User-Agent": "Mozilla/5.0"}


def fetch_daily(sym, days=320):
    """Completed daily candles, ascending [(ts,o,h,l,c)]. Coinbase caps 300/req."""
    now = dt.datetime.now(dt.timezone.utc)
    end = now.replace(hour=0, minute=0, second=0, microsecond=0)  # exclude today (incomplete)
    rows = {}
    t1 = end
    while days > 0:
        take = min(days, 300)
        t0 = t1 - dt.timedelta(days=take)
        u = CB.format(sym=sym, s=t0.isoformat(), e=t1.isoformat())
        data = json.loads(urllib.request.urlopen(
            urllib.request.Request(u, headers=UA), timeout=30).read())
        for ts, lo, hi, op, cl, vol in data:
            if ts < int(end.timestamp()):          # completed bars only
                rows[int(ts)] = (op, hi, lo, cl)
        days -= take
        t1 = t0
    ts_sorted = sorted(rows)
    return ts_sorted, [rows[t] for t in ts_sorted]


def sma(x, w, i):
    return sum(x[i - w + 1:i + 1]) / w if i >= w - 1 else None


def ema_series(x, span):
    k = 2.0 / (span + 1)
    out, e = [], None
    for v in x:
        e = v if e is None else v * k + e * (1 - k)
        out.append(e)
    return out


def atr14(h, l, c, i):
    if i < 14:
        return None
    trs = [max(h[j] - l[j], abs(h[j] - c[j - 1]), abs(l[j] - c[j - 1]))
           for j in range(i - 13, i + 1)]
    return sum(trs) / 14


def evaluate(sym, equity, risk_pct):
    ts, bars = fetch_daily(sym)
    o = [b[0] for b in bars]; h = [b[1] for b in bars]
    l = [b[2] for b in bars]; c = [b[3] for b in bars]
    i = len(c) - 1                                   # last COMPLETED daily bar
    if i < 205:
        return {"sym": sym, "error": f"only {i+1} daily bars; need 206"}
    e9 = ema_series(c, 9)
    s200 = sma(c, 200, i); s50 = sma(c, 50, i); s50p = sma(c, 50, i - 5)
    bear = c[i] < s200
    recovery = bear and c[i] > s50 and s50 > s50p
    regime = "BULL" if not bear else ("RECOVERY" if recovery else "KNIFE")

    # consecutive closes below EMA9 up to yesterday
    below = 0
    j = i - 1
    while j >= 0 and c[j] < e9[j]:
        below += 1; j -= 1
    crossed_up = c[i] > e9[i] and below >= 3
    exit_flag = c[i] < e9[i]

    out = {
        "sym": sym, "regime": regime, "signal": "NONE",
        "asof": dt.datetime.fromtimestamp(ts[i], dt.timezone.utc).strftime("%Y-%m-%d"),
        "close": c[i], "ema9": round(e9[i], 2),
        "sma50": round(s50, 2), "sma200": round(s200, 2),
        "exit_flag": bool(exit_flag),
        "be_arm_pct": 0.02, "be_lock": 0.0,
    }
    if regime == "RECOVERY" and crossed_up:
        a = atr14(h, l, c, i)
        stop = l[i] - 0.5 * (a or 0)
        risk_ps = c[i] - stop
        if risk_ps > 0:
            out.update(signal="ENTER", entry_ref=c[i], stop=round(stop, 2),
                       qty_usd=round(equity * risk_pct / risk_ps * c[i], 2),
                       detail=f"EMA9 reclaim after {below}d below; stop=day-low-0.5*ATR; "
                              f"BE floor arms at +2% MFE; exit first daily close<EMA9")
    elif regime == "RECOVERY":
        out["signal"] = "HOLD-RULES"
        out["detail"] = "in recovery sub-regime; waiting for EMA9 reclaim cross (>=3d below)"
    elif regime == "KNIFE":
        out["detail"] = ("KNIFE phase: FLAT by design -- no long-only entry survives here "
                         "(backtest/crypto_bear_bounce/FINDINGS.md sect.3)")
    else:
        out["detail"] = "BULL regime: Luke daily system owns this phase"
    if exit_flag:
        out["signal"] = "EXIT" if out["signal"] in ("NONE", "HOLD-RULES") else out["signal"]
    return out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--symbols", default="BTC-USD,ETH-USD")
    ap.add_argument("--equity", type=float, default=100_000.0)
    ap.add_argument("--risk-pct", type=float, default=0.02,
                    help="2%% ship tier; 3%% max-aggressive (FINDINGS.md sect.6)")
    a = ap.parse_args()
    for sym in a.symbols.split(","):
        try:
            print(json.dumps(evaluate(sym.strip(), a.equity, a.risk_pct)))
        except Exception as e:
            print(json.dumps({"sym": sym.strip(), "error": str(e)}))


if __name__ == "__main__":
    main()

File: tools/crypto_bear_recovery.cpp
// crypto_bear_recovery.cpp -- BearRecovery signal CLI (C++, replaces the
// earlier Python emitter per operator rule "no python in Omega").
//
// Thin shell around include/CryptoBearRecoveryEngine.hpp: feed it daily
// candles on stdin, it prints one JSON line with the regime-ladder state and
// any ENTER/EXIT signal for the last COMPLETED UTC daily bar. Consumed by the
// ~/Crypto book / omega_crypto_bridge.py (paper 4002) or read manually.
//
// Build:  g++ -O2 -std=c++17 -Iinclude -o /tmp/cbr tools/crypto_bear_recovery.cpp
// Usage (Coinbase raw JSON, newest-first, 300 daily candles -- no key needed):
//   curl -s "https://api.exchange.coinbase.com/products/BTC-USD/candles?granularity=86400" \
//     | /tmp/cbr BTC-USD
//   /tmp/cbr BTC-USD --equity 100000 --risk-pct 0.02 < daily.csv   # ts,o,h,l,c also accepted
//
// Output:
//   {"sym":"BTC-USD","regime":"KNIFE","signal":"NONE","asof":"2026-07-01",
//    "close":...,"ema9":...,"sma50":...,"sma200":...,"stop":...,"qty_usd":...,
//    "be_arm_pct":0.02,"be_lock":0.0,"exit_flag":true,"detail":"..."}
// signal: ENTER (reclaim cross fired) | EXIT (close<EMA9) | HOLD-RULES | NONE.
// Protection semantics + backtested verdict: see the engine header
// (ADVERSE-PROTECTION block) and backtest/crypto_bear_bounce/FINDINGS.md.
#include "../include/CryptoBearRecoveryEngine.hpp"
#include <algorithm>
#include <cctype>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <string>
#include <vector>

struct Bar { long long ts; double o, h, l, c; };

// Coinbase candle JSON: [[time,low,high,open,close,volume],...] newest-first.
// Minimal parser: pull numbers, group by 6.
static std::vector<Bar> parse_coinbase_json(const std::string& s) {
    std::vector<double> nums;
    const char* p = s.c_str();
    while (*p) {
        if (std::isdigit((unsigned char)*p) || (*p == '-' && std::isdigit((unsigned char)p[1]))) {
            char* end = nullptr;
            nums.push_back(std::strtod(p, &end));
            p = end;
        } else ++p;
    }
    std::vector<Bar> out;
    for (size_t i = 0; i + 5 < nums.size(); i += 6) {
        Bar b; b.ts = (long long)nums[i];
        b.l = nums[i + 1]; b.h = nums[i + 2]; b.o = nums[i + 3]; b.c = nums[i + 4];
        out.push_back(b);
    }
    return out;
}

static std::vector<Bar> parse_csv(const std::string& s) {
    std::vector<Bar> out;
    size_t pos = 0;
    while (pos < s.size()) {
        size_t nl = s.find('\n', pos);
        if (nl == std::string::npos) nl = s.size();
        Bar b;
        if (std::sscanf(s.c_str() + pos, "%lld,%lf,%lf,%lf,%lf", &b.ts, &b.o, &b.h, &b.l, &b.c) == 5)
            out.push_back(b);
        pos = nl + 1;
    }
    return out;
}

int main(int argc, char** argv) {
    std::string sym = argc > 1 ? argv[1] : "BTC-USD";
    double equity = 100000.0, risk = 0.02;
    for (int i = 2; i + 1 < argc; i += 2) {
        if (!std::strcmp(argv[i], "--equity"))   equity = std::atof(argv[i + 1]);
        if (!std::strcmp(argv[i], "--risk-pct")) risk   = std::atof(argv[i + 1]);
    }

    std::string in;
    char buf[65536]; size_t n;
    while ((n = std::fread(buf, 1, sizeof buf, stdin)) > 0) in.append(buf, n);
    const size_t first = in.find_first_not_of(" \t\r\n");
    if (first == std::string::npos) { std::printf("{\"sym\":\"%s\",\"error\":\"empty stdin\"}\n", sym.c_str()); return 1; }
    auto bars = in[first] == '[' ? parse_coinbase_json(in) : parse_csv(in);
    std::sort(bars.begin(), bars.end(), [](const Bar& a, const Bar& b) { return a.ts < b.ts; });
    bars.erase(std::unique(bars.begin(), bars.end(),
                           [](const Bar& a, const Bar& b) { return a.ts == b.ts; }), bars.end());

    // drop the incomplete current UTC day
    const long long today0 = (long long)(time(nullptr) / 86400) * 86400;
    while (!bars.empty() && bars.back().ts >= today0) bars.pop_back();
    if (bars.size() < 206) {
        std::printf("{\"sym\":\"%s\",\"error\":\"only %zu completed daily bars; need 206\"}\n",
                    sym.c_str(), bars.size());
        return 1;
    }

    omega::CryptoBearRecoveryEngine eng;
    eng.symbol = sym; eng.equity_ref = equity; eng.risk_pct = risk;
    for (const auto& b : bars) eng.seed_bar(b.o, b.h, b.l, b.c, b.ts * 1000LL);
    const auto st = eng.state();

    const char* signal = "NONE"; const char* detail;
    if (!std::strcmp(st.regime, "RECOVERY") && st.enter_signal) {
        signal = "ENTER";
        detail = "EMA9 reclaim in recovery sub-regime; stop=day-low-0.5*ATR; "
                 "BE floor arms at +2% MFE; exit first daily close<EMA9";
    } else if (!std::strcmp(st.regime, "RECOVERY")) {
        signal = st.exit_flag ? "EXIT" : "HOLD-RULES";
        detail = "recovery sub-regime; waiting for EMA9 reclaim cross (>=3d below)";
    } else if (!std::strcmp(st.regime, "KNIFE")) {
        signal = st.exit_flag ? "EXIT" : "NONE";
        detail = "KNIFE phase: FLAT by design -- no long-only entry survives here "
                 "(backtest/crypto_bear_bounce/FINDINGS.md sect.3)";
    } else if (!std::strcmp(st.regime, "BULL")) {
        detail = "BULL regime: Luke daily system owns this phase";
    } else {
        detail = "warming up (needs 206 completed daily bars)";
    }

    time_t bt = (time_t)(st.bar_ts_ms / 1000);
    struct tm g; gmtime_r(&bt, &g);
    std::printf("{\"sym\":\"%s\",\"regime\":\"%s\",\"signal\":\"%s\","
                "\"asof\":\"%04d-%02d-%02d\",\"close\":%.2f,\"ema9\":%.2f,"
                "\"sma50\":%.2f,\"sma200\":%.2f,\"stop\":%.2f,\"qty_usd\":%.2f,"
                "\"be_arm_pct\":0.02,\"be_lock\":0.0,\"exit_flag\":%s,\"detail\":\"%s\"}\n",
                sym.c_str(), st.regime, signal,
                g.tm_year + 1900, g.tm_mon + 1, g.tm_mday,
                st.close, st.ema9, st.sma50, st.sma200, st.stop, st.qty_usd,
                st.exit_flag ? "true" : "false", detail);
    return 0;
}
